source/kernel/src/registry/mod.rs

```diff
@@ -0,0 +1,440 @@
+use core::any::TypeId;
+
+use mnemos_alloc::{containers::HeapFixedVec, heap::HeapGuard};
+use postcard::experimental::max_size::MaxSize;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use uuid::{uuid, Uuid};
+
+use crate::comms::{
+    bbq,
+    kchannel::{KProducer, LeakedKProducer},
+    rosc::Sender,
+};
+
+/// A partial list of known UUIDs of driver services
+pub static KNOWN_UUIDS: &[Uuid] = &[
+    //
+    // Kernel UUIDs
+    //
+
+    // SerialMux
+    uuid!("54c983fa-736f-4223-b90d-c4360a308647"),
+    // Simple Serial Port
+    uuid!("f06aac01-2773-4266-8681-583ffe756554"),
+];
+
+/// A marker trait designating a registerable driver service.
+///
+/// Typically used with [Registry::set] or [Registry::set_konly].
+/// Can typically be retrieved by [Registry::get] or [Registry::get_userspace]
+/// After the service has been registered.
+pub trait RegisteredDriver {
+    /// This is the type of the request sent TO the driver service
+    type Request: 'static;
+
+    /// This is the type of the response sent FROM the driver service
+    type Response: 'static;
+
+    /// This is the UUID of the driver service
+    const UUID: Uuid;
+
+    /// Get the type_id used to make sure that driver instances are correctly typed.
+    /// Corresponds to the same type ID as `(Self::Request, Self::Response)`
+    fn type_id() -> TypeId {
+        TypeId::of::<(Self::Request, Self::Response)>()
+    }
+}
+
+/// The driver registry used by the kernel.
+pub struct Registry {
+    items: HeapFixedVec<RegistryItem>,
+}
+
+// TODO: This probably goes into the ABI crate, here is fine for now
+#[derive(Serialize, Deserialize)]
+pub struct UserRequest<'a> {
+    // TODO: Maybe not the UUID, maybe pre-discover a shorter UID?
```

Comment:
we could maybe avoid the linear search of the registry if we had a way for userspace to look up the array index of a driver by its UUID, or something? then, you could do the linear search once for driver discovery, and then reference the driver by an index every time you want to talk to it.

Reply:
Opened https://github.com/tosc-rs/mnemos/issues/32 to track this, I'd like to keep it simple for now and try using the registry before we spend time optimizing it (you make a very good point though!)